contrib/src/main/scala/akka/stream/contrib/FocusedFlow.scala

```diff
@@ -0,0 +1,241 @@
+   graph: Graph[IncompleteFocusedSourceShape[(Tag, In), In, Out], Mat]
+  ) extends FlowOpsMat[Out, Mat] {
+  import GraphDSL.Implicits._
+
+  override type Repr[+O] = FocusedSource[Tag, In, O, Mat@uncheckedVariance]
+  override type ReprMat[+O, +M] = FocusedSource[Tag, In, O, M]
+  override type Closed = RunnableGraph[Mat@uncheckedVariance]
+  override type ClosedMat[+M] = RunnableGraph[M]
+
+  def unfocus: Source[(Tag, Out), Mat] = Source.fromGraph(
+    GraphDSL.create(graph) {
+      implicit builder =>
+      g =>
+        val tagger = builder.add(new FlowTagger[In, Out, Tag])
+
+        g.srcIn ~> tagger.in1
+        tagger.out2 ~> g.in
+        g.out ~> tagger.in2
+
+        SourceShape(tagger.out1)
+    })
+
+  override def via[T, Mat2](flow: Graph[FlowShape[Out, T], Mat2]): FocusedSource[Tag, In, T, Mat] =
+    viaMat(flow)(Keep.left)
+  override def viaMat[T, Mat2, Mat3](flow: Graph[FlowShape[Out, T], Mat2])(combine: (Mat, Mat2) => Mat3): FocusedSource[Tag, In, T, Mat3] =
+    new FocusedSource(GraphDSL.create(graph, flow)(combine) {
+      implicit builder =>
+        (g, f) =>
+          g.out ~> f
+          IncompleteFocusedSourceShape(g.srcIn, g.in, f.out)
+    })
+
+  // connecting the focused flow to a sink will discard the tag
+  override def to[Mat2](sink: Graph[SinkShape[Out], Mat2]): RunnableGraph[Mat] =
+  toMat(sink)(Keep.left)
+  override def toMat[Mat2, Mat3](sink: Graph[SinkShape[Out], Mat2])(combine: (Mat, Mat2) => Mat3): RunnableGraph[Mat3] =
+    RunnableGraph.fromGraph(
+      GraphDSL.create(graph, sink)(combine) {
+        implicit builder =>
+          (g, s) =>
+            val mapper = builder.add(new akka.stream.impl.fusing.Map[(_, In), In](_._2))
+            g.srcIn ~> mapper ~> g.in
+            g.out ~> s.in
+            ClosedShape
+      })
+
+  override def mapMaterializedValue[Mat2](f: (Mat) => Mat2): FocusedSource[Tag, In, Out, Mat2] =
+    viaMat(Flow[Out]) { case (m, notused) => f(m) }
+
+  override def withAttributes(attr: Attributes): FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.withAttributes(attr))
+
+  override def addAttributes(attr: Attributes): FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.addAttributes(attr))
+
+  override def named(name: String): FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.named(name))
+
+  override def async: FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.async)
+}
+
+private class FlowTagger[In, Out, E]
+  extends GraphStage[BidiShape[(E, In), (E, Out), Out, In]] {
+  val inTup = Inlet[(E, In)]("in-tuple")
+  val outTup = Outlet[(E, Out)]("out-tuple")
+  val toFlow = Outlet[In]("to-flow")
+  val fromFlow = Inlet[Out]("from-flow")
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    var curr: E = _
+    setHandler(inTup, new InHandler {
+      override def onPush(): Unit = {
+        val (cur, in) = grab(inTup)
+        curr = cur
```

Comment:
@ktoso
So this brings up a couple of related questions:
- If the user has to provide the 'get' and 'set' mappings, then why limit it to tuples?
- If it is limited to tuples, then wouldn't it be better to generate methods like focus_1, focus_2,... ?
- What happens if the output of the flow is a different type from the input?

Originally I thought about having the user to provide the functions when calling focus/unfocus, but that made the API really cluttered and I don't think it saved too much (considering that the user can just add a 'map' themselves).

So what I'm trying to say is, I'll gladly try to make it work for larger tuples, but I'm not sure whether it really adds much.